### PR TITLE
Move flags from UserOverride.cmake to CMakeLists.txt

### PR DESCRIPTION
**Pull request by isuruf**
Fixes https://github.com/symengine/symengine.py/issues/206
Fixes https://github.com/symengine/symengine/issues/1422

Some build systems (like Homebrew) set flags like `CMAKE_CXX_FLAGS_Release` to different defaults than CMake, but doesn't add `-std=c++11` and therefore builds wouldn't have the c++11 flag added.

This PR uses the CMake defaults and moves our custom flags to `CMakeLists.txt` which will be used regardless of whether `CMAKE_CXX_FLAGS_Release` is set or not.

TODO:
- [x] SymEngineConfig.cmake.in

### CODE REVIEW

File: CMakeLists.txt [hunk 1]
```diff
@@ -12,38 +10,52 @@ set(CMAKE_MODULE_PATH ${PROJECT_SOURCE_DIR}/cmake)
...
 endif()
 
 if (CMAKE_CXX_COMPILER_ID MATCHES "GNU|Clang")
-    set(CMAKE_CXX_FLAGS_DEBUG
-        "${CMAKE_CXX_FLAGS_DEBUG} -Wno-unused-parameter -ggdb")
-    set(CMAKE_CXX_FLAGS_RELEASE
-        "${CMAKE_CXX_FLAGS_RELEASE} -Wno-unused-parameter")
-    if (NOT BUILD_FOR_DISTRIBUTION)
-        set(CMAKE_CXX_FLAGS_RELEASE
-            "${CMAKE_CXX_FLAGS_RELEASE} -march=native")
```
Comment (isuruf) [inline CMakeLists.txt]: I've removed this and then checked the benchmarks and there were no noticeable difference. @certik, can you check?

Comment (certik) [reply CMakeLists.txt]: The `march=native` makes a big difference on some computers. I would leave it there for now.

**Comment by isuruf** [comment]
Needs a review

**Comment by certik** [comment]
Hm, github didn't post my comment from email. 

How did the benchmark change with this change? We don't need to use the UserOverride.cmake, but we need some way to supply the correct compiler options by default.

**Comment by isuruf** [comment]
> How did the benchmark change with this change?

If you use the same options as before, no change in the flags, therefore no change in benchmarks. `cmake -DCMAKE_BUILD_TYPE=Release/Debug -DBUILD_FOR_DISTRIBUTION=on/off`

> We don't need to use the UserOverride.cmake, but we need some way to supply the correct compiler options by default.

`cmake -DCMAKE_CXX_FLAGS_Release=-O2` is a way to override the default compiler flags by cmake.
`CXXFLAGS='-O3' cmake .` is a way to append to the flags.
These are the 2 standard ways to override and append compiler flags.

**Comment by certik** [comment]
In current master, if you do `cmake .`, you get:
```
CMAKE_BUILD_TYPE: Release
CMAKE_CXX_FLAGS_DEBUG:   -std=c++11 -Wall -Wextra -fno-common -fPIC -g -Wno-unused-parameter -ggdb -Wno-unknown-pragmas
CMAKE_CXX_FLAGS_RELEASE: -std=c++11 -Wall -Wextra -fno-common -fPIC -O3 -ffast-math -funroll-loops -Wno-unused-parameter -march=native -Wno-unknown-pragmas
```
and with this PR, you get:
```
CMAKE_BUILD_TYPE: RelWithDebInfo
CMAKE_CXX_FLAGS:                   -fPIC -std=c++11 -Wall -Wextra -fno-common   -Wno-unused-parameter -Wno-unknown-pragmas
CMAKE_CXX_FLAGS_DEBUG:             -g
CMAKE_CXX_FLAGS_RELEASE:           -O3 -DNDEBUG
CMAKE_CXX_FLAGS_RELWITHDEBINFO:    -O2 -g -DNDEBUG
```
So the default options changed from `-std=c++11 -Wall -Wextra -fno-common -fPIC -O3 -ffast-math -funroll-loops -Wno-unused-parameter -march=native -Wno-unknown-pragmas` to ` -O2 -g -DNDEBUG`. I think that's a big problem.

**Comment by isuruf** [comment]
@certik, you have to look at `CMAKE_CXX_FLAGS + CMAKE_CXX_FLAGS_<BUILD_TYPE>`. Both of these are used in BUILD_TYPE.

Should we keep `Release` as the default or `RelwithDebInfo`

**Comment by certik** [comment]
@isuruf I might be misunderstanding. I am just trying to figure it out. I posted what I saw in the output [above](https://github.com/symengine/symengine/pull/1376#issuecomment-377581107). Even if you append `CMAKE_CXX_FLAGS` and `CMAKE_CXX_FLAGS_RELEASE`, you get: `-fPIC -std=c++11 -Wall -Wextra -fno-common   -Wno-unused-parameter -Wno-unknown-pragmas -O3 -DNDEBUG` which is different to the current master, which has the following extra options: `-ffast-math -funroll-loops -march=native`.

**Comment by isuruf** [comment]
@certik, you are right. `-march=native` is added now with the last commit. I'll add `-funroll-loops` as well.

I don't really like `-ffast-math` because you can't work with `nan`s and `inf`s if you use that flag. `-ffast-math` tells the compile that a `double` or a `float` can never be `nan` or `inf`s.

**Comment by certik** [comment]
We can discuss and  remove `-ffast-math` in a separate PR. In here, let's just exactly replicate all options from master.

What is the hash of the "last commit"? I can see this one as the last commit: https://github.com/symengine/symengine/pull/1376/commits/97ce774da568f2de70cfa2416db2ae800d37a899, which does not add `-march=native`.

**Comment by isuruf** [comment]
It's the correct commit. Here's my output of 97ce774
```
CMAKE_BUILD_TYPE: RelWithDebInfo
CMAKE_CONFIGURATION_TYPES: 
CMAKE_CXX_FLAGS:                   -fPIC -std=c++11 -Wall -Wextra -fno-common   -Wno-unused-parameter -Wno-unknown-pragmas
CMAKE_CXX_FLAGS_DEBUG:             -g
CMAKE_CXX_FLAGS_RELEASE:           -O3 -DNDEBUG -march=native
CMAKE_CXX_FLAGS_RELWITHDEBINFO:    -O2 -g -DNDEBUG
```

**Comment by certik** [comment]
I see, I get the same output. So let's add the other two options `-ffast-math -funroll-loops` and also switch from `RelWithDebInfo` to `Release`. I'll then review it more.

**Comment by certik** [comment]
Ok, now by default I am getting the correct options. However, now changing the options does not work.

This correctly appends an option:
```
$ CXXFLAGS='-O2' cmake .
...
CMAKE_BUILD_TYPE: Release
CMAKE_CONFIGURATION_TYPES: 
CMAKE_CXX_FLAGS:                   -fPIC -std=c++11 -Wall -Wextra -fno-common -O
2   -Wno-unused-parameter -funroll-loops -ffast-math -Wno-unknown-pragmas
CMAKE_CXX_FLAGS_DEBUG:             -g
CMAKE_CXX_FLAGS_RELEASE:           -O3 -DNDEBUG -march=native
CMAKE_CXX_FLAGS_RELWITHDEBINFO:    -O2 -g -DNDEBUG
```
But this does not replace the option:
```
$ cmake -DCMAKE_CXX_FLAGS_RELEASE=-O2 .
...
CMAKE_BUILD_TYPE: Release
CMAKE_CONFIGURATION_TYPES: 
CMAKE_CXX_FLAGS:                   -fPIC -std=c++11 -Wall -Wextra -fno-common   -Wno-unused-parameter -funroll-loops -ffast-math -Wno-unknown-pragmas
CMAKE_CXX_FLAGS_DEBUG:             -g
CMAKE_CXX_FLAGS_RELEASE:           -O2 -march=native
CMAKE_CXX_FLAGS_RELWITHDEBINFO:    -O2 -g -DNDEBUG
```
As you can see, now the user is stuck with the `-march=native` option and no way to remove it. What is worse, this seems to not have any effect: `-DCMAKE_CXX_FLAGS=-O2`.

We need to allow the user to completely specify compiler options. Previously this was possible, but now it isn't. For example the user might want to *not* to use any of the `-march=native -funroll-loops -ffast-math` options.
```

**Comment by certik** [comment]
@isuruf it might not be possible to do what I want, that's why I settled on using UserOverride.cmake. At the end of the day, I only have two requirements:

1. By default, `cmake .` should give the correct optimizing options for the compiler selected by the user using the CC variable.

2. It must be possible for the user to completely override the default options and set any options he wants.

We can sacrifice "appending" options to our default options, we can also sacrifice other things. But I feel the above requirements 1. and 2. are extremely important to have. 1. is needed so that SymEngine is performing for anybody who installs it, so that they don't need to read the README and painfully try to figure out what the correct options are. 2. is needed so that if SymEngine fails, the user has an easy way to override the default options and do whatever he wants for his particular case.

As it is now, this PR seems to fail 2., which means the user has no way to remove the `-march=native -funroll-loops -ffast-math` options, short of patching our `CMakeLists.txt` file. So we need to fix this.

**Comment by certik** [comment]
Regarding the brew issue, it seems it's a bug in brew if it sets `CMAKE_CXX_FLAGS_RELEASE` and does not enable the correct options for C++11. So perhaps the correct option should be added in the brew package?

**Comment by isuruf** [comment]
> Regarding the brew issue, it seems it's a bug in brew if it sets CMAKE_CXX_FLAGS_RELEASE and does not enable the correct options for C++11

In homebrew, these flags are default ones for all homebrew cmake packages. Some need C++11, some need C++14, etc. C++11 is a requirement for symengine and symengine should set it no matter the flags set by the user. Otherwise user would have to know what flags to set.

**Comment by isuruf** [comment]
@certik can you check again?

**Comment by certik** [comment]
In Homebrew, how are cmake packages supposed to set the correct flags in addition to the default homebrew flags? Also, because we want users to have the fastest build of symengine in homebrew, the homebrew symengine package should set the correct flags anyway (or let symengine use its default).

**Comment by isuruf** [comment]
> In Homebrew, how are cmake packages supposed to set the correct flags in addition to the default homebrew flags?

They should be added to the CMakeLists.txt not UserOverride.cmake.

> Also, because we want users to have the fastest build of symengine in homebrew, the homebrew symengine package should set the correct flags anyway.

They are making bottles (binary packages), so architecture specific optimizations should be turned off anyways.

**Comment by certik** [comment]
I checked the master, and the master branch is already broken regarding setting the flags. :(

This PR doesn't fix it either, because it sets options in CMakeLists.txt, so users literally have no way to override them if they wanted. 

This is a defficiency of CMake, and UserOverride is the only way I figured out to make it satisfy 1. and 2. above.

Right now, both master and this PR fails 2. That's a problem.

Regarding homebrew, let's simply make the homebrew package append `-std=c++11` to it's CMAKE_CXX_FLAGS_RELEASE.

**Comment by certik** [comment]
Where is the homebrew symengine formula? I can't find it. Let me have a look there and see what is going on. We should test this on travis to ensure things work.

**Comment by isuruf** [comment]
> Right now, both master and this PR fails 2. That's a problem.

Which flags can't be completely removed in master and this PR?

**Comment by certik** [comment]
In master I can't remove `-march=native`, ` -Wno-unused-parameter` and `-Wno-unknown-pragmas`:
```
$ cmake -DCMAKE_CXX_FLAGS_RELEASE="-O3 -std=c++11" .
...
CMAKE_CXX_FLAGS_DEBUG:   -std=c++11 -Wall -Wextra -fno-common -fPIC -g -Wno-unused-parameter -ggdb -Wno-unknown-pragmas
CMAKE_CXX_FLAGS_RELEASE: -O3 -std=c++11 -Wno-unused-parameter -march=native -Wno-unknown-pragmas
...
```

**Comment by certik** [comment]
We should have a test for this ---- test that 1. and 2. works. I really don't care *how* it is implemented and whether we use UserOverride, or how the user can override the flags. But I very much are that both 1. and 2. works.

Probably the best way to test it is to create a simple compiler wrapper script and check that the compiler is provided with the expected flags both in 1. and 2., and then just test it on Travis.

**Comment by isuruf** [comment]
march=native can be removed by doing `cmake -DBUILD_FOR_DISTRIBUTION=yes`. Warning flags are okay and if the user wants to remove them they can add `-Wunused-parameter` etc, but we can have another cmake option to remove them also.

**Comment by bjodah** [comment]
Sorry if this is slightly off topic, but it has to do with architecture flags: some time back I tried to compile SymEngine with `-mx32` (to see if 32-bit pointers would be beneficial) but I couldn't get it to work, I've been meaning to look into it again but haven't found the time. I'm still rather new to CMake so not sure how to set it appropriately.

Thought I would just mention it as a possible use-case where target architecture is a bit special even though one would not necessarily be building for distribution.

**Comment by certik** [comment]
Closing this in favor of #1431. I hope I fixed everything and I tried to explain in the commit log how it works.

**Pull request closed without merging**